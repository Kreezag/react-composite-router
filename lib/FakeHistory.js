export default class FakeHistory {
  constructor () {
    this.location = {
      pathname: '',
      hash: '',
      search: ''
    };

    this.action = '';

    this._listeners = [];
  }

  listen (listener) {
    this._listeners = this._listeners.concat([ listener ]);

    return () => {
      this._listeners = this._listeners.filter((_listener) => _listener !== listener);
    };
  }

  _fire () {
    this._listeners.forEach((listener) => {
      listener(this.location, this.action);
    });
  }

  replace (location) {
    this.location = Object.assign({}, this.location, location);
    this.action = 'REPLACE';

    this._fire();
  }

  push (location) {
    this.location = Object.assign({}, this.location, location);
    this.action = 'PUSH';

    this._fire();
  }

  createHref (location) {
    return `${location.pathname}${location.search ? `?${location.search}` : ''}${location.hash ? `#${location.hash}` : ''}`;
  }
}
