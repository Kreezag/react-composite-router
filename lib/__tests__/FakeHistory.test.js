/* eslint-env browser, jest */

import FakeHistory from '../FakeHistory';


describe('FakeHistory', () => {

  it('should run listeners queue', () => {
    const history = new FakeHistory();

    const listener1 = jest.fn();
    const listener2 = jest.fn();

    history.listen(listener1);

    history.push({ pathname: '', hash: '', search: '' });

    const destroy = history.listen(listener2);

    destroy();

    history.replace({ pathname: '', hash: '', search: '' });

    expect(listener1).toBeCalled();
    expect(listener2).not.toBeCalled();
  });

  it('calc href', () => {
    const history = new FakeHistory();

    expect(history.createHref({ pathname: '/', hash: 'some-hash', search: 'some=search' })).toMatchSnapshot();
    expect(history.createHref({ pathname: '/', hash: 'some-hash', search: '' })).toMatchSnapshot();
  });
});
