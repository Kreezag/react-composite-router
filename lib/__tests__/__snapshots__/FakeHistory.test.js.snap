exports[`FakeHistory calc href 1`] = `"/?some=search#some-hash"`;

exports[`FakeHistory calc href 2`] = `"/#some-hash"`;
